File: ios/Sources/WMBridgeCore/WMBridge.swift
import Foundation

/// Common contract implemented by exactly one of `WMStandardAdapter` (backed
/// by `WalkMeSDK` from the `WalkMe` product) or `WMEditorAdapter` (backed by
/// `WalkMePowerMode` from the `WalkMeEditor` product) — whichever adapter
/// target Package.swift links in for the chosen `walkmeVariant`.
public protocol WMBridge: AnyObject {
    var variantName: String { get } // "standard" | "editor"

    func start(options: WMStartOptions)
    func stop()
    func restart()

    func setUserId(_ userId: String?)
    func setLanguage(_ language: String)
    func setVariable(key: String, value: String?)
    func setEventUserVars(_ values: [String: String])
    func setTenantId(_ tenantId: String?)

    func startItem(byID itemId: String, deepLink: String?)
    func dismissItem()

    func sendEvent(name: String, attributes: [String: String]?)

    /// Registers native callbacks; adapters forward these to the emitter.
    func setEventEmitter(_ emitter: WMEventEmitter)
}

public struct WMStartOptions {
    public let systemGuid: String
    public let environment: String
    public let dataCenter: String // "prod" | "eu" | "us01" | "eu01" | custom string
    public let analyticsEnabled: Bool
    public let localLogsEnabled: Bool
    public let language: String?
    public let userId: String?

    public init(
        systemGuid: String,
        environment: String = "Production",
        dataCenter: String = "prod",
        analyticsEnabled: Bool = true,
        localLogsEnabled: Bool = false,
        language: String? = nil,
        userId: String? = nil
    ) {
        self.systemGuid = systemGuid
        self.environment = environment
        self.dataCenter = dataCenter
        self.analyticsEnabled = analyticsEnabled
        self.localLogsEnabled = localLogsEnabled
        self.language = language
        self.userId = userId
    }
}

public protocol WMEventEmitter: AnyObject {
    func onItemPresented(itemId: String, itemType: String?, userData: [String: Any]?)
    func onItemDismissed(itemId: String, actionType: String?, userData: [String: Any]?)
    func onItemAction(itemId: String, args: [String: String]?)
    func onAnalyticsEvent(eventName: String, params: [String: Any])
}

File: android/src/main/kotlin/com/walkme/capacitor/WMBridge.kt
package com.walkme.capacitor

import android.app.Activity

/**
 * Common contract implemented by exactly one of [WMStandardBridge] (backed by
 * `com.walkme.sdk.WalkMeSDK`) or [WMEditorBridge] (backed by
 * `com.walkme.pm.WalkmeSdkPowerMode`) — whichever one Gradle compiled in for
 * this build's `walkmeVariant`. [WalkMePlugin] talks only to this interface,
 * so it never needs to know which native artifact is actually present.
 */
interface WMBridge {

    val variantName: String // "standard" | "editor"

    fun start(activity: Activity, options: WMStartOptions)
    fun stop()

    fun restart()

    fun setUserId(userId: String?)
    fun setLanguage(language: String)
    fun setVariable(key: String, value: String?)
    fun setEventUserVars(values: Map<String, String>)
    fun setTenantId(tenantId: String?)

    fun startItemByID(itemId: String, deepLink: String?)

    fun dismissItem()

    fun sendEvent(name: String, attributes: Map<String, String>?)

    /** Registers native callbacks; bridge implementations forward these to [WMEventEmitter]. */
    fun setEventEmitter(emitter: WMEventEmitter)
}

/** Startup options shared across variants (mirrors `WalkMeStartOptions` on both native SDKs). */
data class WMStartOptions(
    val systemGuid: String,
    val environment: String = "Production",
    val dataCenter: String = "prod",
    val analyticsEnabled: Boolean = true,
    val localLogsEnabled: Boolean = false,
    val language: String? = null,
    val userId: String? = null,
)

/** Sink the bridge implementations push native lifecycle/analytics events into. */
interface WMEventEmitter {
    fun onItemPresented(itemId: String, itemType: String?, userData: Map<String, Any?>?)
    fun onItemDismissed(itemId: String, actionType: String?, userData: Map<String, Any?>?)
    fun onItemAction(itemId: String, args: Map<String, String>?)
    fun onAnalyticsEvent(eventName: String, params: Map<String, Any?>)
}
